crypto/options.go

```diff
@@ -13,17 +23,108 @@ func CollectSigningOptions(opts []SigningOption) SigningOpts {
+	if opts.payloadEncoding == 0 {
+		return varsig.PayloadEncodingVerbatim
+	}
+	return opts.payloadEncoding
 }
 
+func (opts SigningOpts) VarsigMatch(disc varsig.Discriminator, curve uint64, keyLength uint64) bool {
+	// This is relatively ugly, but we get cyclic import otherwise
+	switch opts.disc {
+	case 0:
+		// not varsig to compare
+		return true
+	case varsig.DiscriminatorECDSA:
+		return disc == varsig.DiscriminatorECDSA && opts.curve == curve
+	case varsig.DiscriminatorEdDSA:
+		return disc == varsig.DiscriminatorEdDSA && opts.curve == curve
+	case varsig.DiscriminatorRSA:
+		return disc == varsig.DiscriminatorRSA && opts.keyLen == keyLength
+	default:
+		panic("unreachable")
+	}
+}
+
+// // Varsig return a non-nil varsig.Varsig if WithVarsig has been used.
+// func (opts SigningOpts) Varsig() varsig.Varsig {
+// 	return opts.varsig
+// }
+
+// func (opts SigningOpts) VarsigMatch(priv PrivateKey) bool {
+// 	if opts.varsig == nil {
+// 		return false
+// 	}
+//
+// 	switch vsig := opts.varsig.(type) {
+// 	case varsig.ECDSAVarsig:
+// 		switch priv.(type) {
+// 		case *p256.PrivateKey:
+// 			return vsig.Curve() == varsig.CurveP256
+// 		case *p384.PrivateKey:
+// 			return vsig.Curve() == varsig.CurveP384
+// 		case *p521.PrivateKey:
+// 			return vsig.Curve() == varsig.CurveP521
+// 		case *secp256k1.PrivateKey:
+// 			return vsig.Curve() == varsig.CurveSecp256k1
+// 		default:
+// 			return false
+// 		}
+// 	case varsig.EdDSAVarsig:
+// 		_, ok := priv.(ed25519.PrivateKey)
+// 		return ok
+// 	case varsig.RSAVarsig:
+// 		switch priv := priv.(type) {
+// 		case *rsa.PrivateKey:
+// 			return vsig.KeyLength() == priv.KeyLength()
+// 		default:
+// 			return false
+// 		}
+// 	default:
+// 		panic("unreachable")
+// 	}
+// }
+
 type SigningOption func(opts *SigningOpts)
 
 // WithSigningHash specify the hash algorithm to be used for signatures
 func WithSigningHash(hash Hash) SigningOption {
 	return func(opts *SigningOpts) {
-		opts.Hash = hash
+		opts.hash = hash
+	}
+}
+
+// WithPayloadEncoding specify the encoding that was used on the message before signing it.
+// This will be included in the resulting varsig.
+func WithPayloadEncoding(encoding varsig.PayloadEncoding) SigningOption {
+	return func(opts *SigningOpts) {
+		opts.payloadEncoding = encoding
+	}
+}
+
+// WithVarsig configure the signing or verification parameters from a varsig.
+// If you use WithVarsig, you should NOT use other options.
+func WithVarsig(vsig varsig.Varsig) SigningOption {
+	return func(opts *SigningOpts) {
+		opts.payloadEncoding = vsig.PayloadEncoding()
+		opts.hash = FromVarsigHash(vsig.Hash())
+		opts.disc = vsig.Discriminator()
+		switch vsig := vsig.(type) {
+		case varsig.EdDSAVarsig:
+			opts.curve = uint64(vsig.Curve())
+		case varsig.ECDSAVarsig:
+			opts.curve = uint64(vsig.Curve())
+		case varsig.RSAVarsig:
+			opts.keyLen = vsig.KeyLength()
+		default:
+			panic("unreachable")
```

Comment:
Shouldn't this return an error ... it may not be used only in an application's initialization.

Reply:
Yeah that's a down side of the design, but also the minimal down side I could find. Other ways are worse or way worse.

Error or panic is not going to change much: in both case that code is not able to handle a new kind of varsig. I know that technically go-varsig can register extension, but would that really happen?